src/vision/submitted_stories/status_quo/grace_debugs_a_crash_dump.md

```diff
@@ -0,0 +1,133 @@
+
+  5 std::thread::local::LocalKey<…>::try_with<…>(…) • thread/local.rs:262
+      self = (*)0x3816da0c9b0 ➔ LocalKey<core::cell::RefCell<core:…>{inner: &distridata_async::executor::EXECUTOR::__getit}
+      f = $(closure-0)($(closure-0)((*)0x1aea32022a0))
+
+  6 std::thread::local::LocalKey<…>::with<…>(…) + 0x27 (no line info)
+      self = (*)0x3816da0c9b0 ➔ LocalKey<core::cell::RefCell<core:…>{inner: &distridata_async::executor::EXECUTOR::__getit}
+      f = $(closure-0)($(closure-0)((*)0x1aea32022a0))
+
+  7 distridata_async::executor::with_local_timer_heap<…>(…) + 0x2a (no line info)
+      f = $(closure-0)((*)0x1aea32022a0 ➔ (*)0x1aea3202758)
+
+▶ 8 distridata_async::executor::Executor::run_singlethreaded<…>(…) • default/../../src/lib/distridata-async/src/executor.rs:393
+      self = (*)0x1aea3202758 ➔ Executor{inner: (*)0x3f0f481a380, next_packet: …}
+      main_future = GenFuture<generator-0>(Unresumed)
+
+  9 distridata_pkg_testing_lib_test::serve::tests::test_serve_empty() • serve.rs:345
+
+  10 λ(…) • serve.rs:345
+      (*)0x1aea3202b80 ➔ $(closure-0)
+
+  11 core::ops::function::FnOnce::call_once<…>(…) • function.rs:232
+      $(closure-0)
+      <Value has no data.>
+```
+
+The backtrace shows a lot of detail about the executor, but not of this is really relevant to Grace's code. She will have to inspect the executor manually in order to find the information she needs. Frame 8 looks promising, so the finds the local variables there and sees one called `main_future`. Inspecting the code, she sees this has a `pointer` field, which might tell her something about the task that's running. She takes a look:
+
+```
+[dbg] print -t --max-array=2 main_future.pointer
+(std::future::GenFuture<generator-0>*) 0x1aea32022a8 ➔ std::future::GenFuture<generator-0>(
+  (distridata_pkg_testing_lib_test::serve::tests::test_serve_empty::func::generator-0) distridata_pkg_testing_lib_test::serve::tests::test_serve_empty::func::$(generator-0)::Suspend6{
+    packages: alloc::vec::Vec<distridata_pkg_testing_lib_test::repo::PackageEntry>[]
+    bytes: alloc::vec::Vec<u8>[123, 34, ]
+    (alloc::string::String) url: "ht"...
+    also_bytes: alloc::vec::Vec<u8>[123, 34, ]
+    pinned: std::future::GenFuture<generator-0>(
+      distridata_pkg_testing_lib_test::serve::get::$(generator-0){
+        (alloc::string::String) __0: "ht"...
+      }
+    )
+  }
+)
+```
+
+This has some more information, but it is still not as helpful as Grace was hoping for.
+
+Grace quickly realizes her tools are not going to give her as much help as she'd like. She does manage to find the Tokio executor in memory, so she starts reading the Tokio code to understand how tasks are laid out in memory, etc. Even once she finds the list of tasks, she can only see the opaque contents of the closure. It is hard even to track these back to a line number, or to what operating system resource the task is blocked on (IOCP handle, io_uring event, etc.).

```

Comment:
There might be some confusion from the fact that this says Tokio, but the executor frames don't mention tokio at all (instead `distridata_async`).

Reply:
Ah, good idea. I removed the reference to Tokio.